### PR TITLE
feat(connect-cardano): add support for babbage features

### PR DESCRIPTION
**Pull request by gabrielKerekes**

### CODE REVIEW

File: packages/connect/src/api/cardanoSignTransaction.ts [hunk 1]
```diff
@@ -265,6 +289,24 @@ export default class CardanoSignTransaction extends AbstractMethod<
         if (params.signingMode === PROTO.CardanoTxSigningMode.PLUTUS_TRANSACTION) {
             this._ensureFeatureIsSupported('Plutus');
         }
+
+        params.outputsWithData.forEach(({ output, inlineDatum, referenceScript }) => {
```
Comment (davidmisiak) [inline packages/connect/src/api/cardanoSignTransaction.ts:295]: What about merging this with the `params.outputsWithData.forEach(({ output })...` above?

File: packages/connect/src/api/cardano/cardanoUtils.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    typedCall: any,
+    data: string,
+    chunkSize: number,
+    messageType: string,
+) => {
+    let processedSize = 0;
+    while (processedSize < data.length) {
+        const chunk = data.slice(processedSize, processedSize + chunkSize);
+        await typedCall(messageType, 'CardanoTxItemAck', {
+            data: chunk,
```
Comment (davidmisiak) [inline packages/connect/src/api/cardano/cardanoUtils.ts:13]: How does this work? `typedCall` expects a hex-encoded string when sending bytes?

File: packages/connect/src/api/cardano/cardanoOutputs.ts [hunk 3]
```diff
@@ -43,3 +57,40 @@ export const transformOutput = (output: any): OutputWithTokens => {
 
     return result;
 };
+
+export const sendOutput = async (typedCall: any, outputWithData: OutputWithData) => {
+    const MAX_CHUNK_SIZE = 512;
```
Comment (davidmisiak) [inline packages/connect/src/api/cardano/cardanoOutputs.ts]: I don't like this number being different than the number in FW (even though it represents the same number of bytes). I would prefer converting the data from hex-encoded string to bytes asap and dealing only with bytes later on - is that possible?

File: packages/connect/src/api/cardano/cardanoOutputs.ts [hunk 4]
```diff
@@ -1,30 +1,44 @@
 // origin: https://github.com/trezor/connect/blob/develop/src/js/core/methods/helpers/cardanoOutputs.js
 
+/* eslint-disable no-restricted-syntax */
```
Comment (davidmisiak) [inline packages/connect/src/api/cardano/cardanoOutputs.ts:4]: Just out of curiosity, why is this needed?

File: packages/connect/src/types/api/cardanoSignTransaction.ts [hunk 5]
```diff
@@ -19,19 +19,21 @@ export interface CardanoAssetGroup {
     tokenAmounts: CardanoToken[];
 }
 
-export type CardanoOutput =
+export type CardanoOutput = (
```
Comment (davidmisiak) [inline packages/connect/src/types/api/cardanoSignTransaction.ts:22]: Hm, it seems like they removed Flow types (finally!), is that right? In that case, we should update/remove the places where we reference them, such as `cardanoSignTransaction.md` and other `.md` files.

Also, `cardanoSignTransaction.md` should be updated with the new Babbage tx items.

File: packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts [hunk 6]
```diff
@@ -347,6 +399,12 @@ const legacyResults = {
         rules: ['<2.4.4', '1'],
         payload: false,
     },
+    beforeBabbage: {
+        // older FW doesn't support babbage-related features (inlineDatum, referenceScript, collateral return,
```
Comment (davidmisiak) [inline packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts]: Nit: `collateral return` has different formatting.

File: packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts [hunk 7]
```diff
@@ -347,6 +399,12 @@ const legacyResults = {
...
+        // totalCollateral, referenceInputs)
+        rules: ['<2.5.2', '1'],
+        payload: false,
+    },
```
Comment (davidmisiak) [inline packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts:417]: Reminder: We should update `multisigWithMostElementsFilledAndSharedWithLedger` (at least add a Babbage-formatted output, or maybe create a similar testcase for Alonzo&Babbage items?).

File: packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts [hunk 8]
```diff
@@ -2101,5 +2166,342 @@ export default {
...
+        },
+
+        {
+            description: 'ordinaryTransactionWithBothOutputFormats',
+            params: {
+                inputs: [SAMPLE_INPUTS.shelley_input],
+                outputs: [
+                    SAMPLE_OUTPUTS.simple_shelley_output,
+                    SAMPLE_OUTPUTS.output_with_datum_hash_and_base_address_paramaters,
+                ],
```
Comment (davidmisiak) [inline packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts]: I'm afraid both outputs use the legacy format (despite the testcase name).

**Review by davidmisiak — COMMENTED** [review]
A few questions and suggestions below. I haven't run anything since I don't have the Suite set up.

File: packages/connect/src/api/cardanoSignTransaction.ts (hunk 1 above)
Comment (gabrielKerekes) [reply packages/connect/src/api/cardanoSignTransaction.ts:295]: I prefer having the checks separated. I like the fact that the checks are chronological as they were added and it's IMHO clearer what's supported with each version.

File: packages/connect/src/api/cardano/cardanoUtils.ts (hunk 2 above)
Comment (gabrielKerekes) [reply packages/connect/src/api/cardano/cardanoUtils.ts:13]: I think so, yes. Similarly to other `bytes` types e.g. `CardanoTxInput.prev_hash` or `CardanoTxOutput.datum_hash`. I think the protobuf layer somehow deals with the conversion to bytes, but I haven't checked exactly. I just know that this worked.

Even when a `Buffer` is used as an incoming type into Connect it gets converted to hex string before being sent to Trezor: https://github.com/trezor/trezor-suite/blob/develop/packages/connect/src/api/bitcoin/inputs.ts#L73

File: packages/connect/src/api/cardano/cardanoOutputs.ts (hunk 3 above)
Comment (gabrielKerekes) [reply packages/connect/src/api/cardano/cardanoOutputs.ts]: I wouldn't convert to bytes before sending, see [comment](https://github.com/vacuumlabs/trezor-suite/pull/1#discussion_r907136861) above.

Perhaps this would be better?
```suggestion
    const MAX_CHUNK_SIZE = 256 * 2; // 256 hex-encoded bytes
```

File: packages/connect/src/api/cardano/cardanoOutputs.ts (hunk 4 above)
Comment (gabrielKerekes) [reply packages/connect/src/api/cardano/cardanoOutputs.ts:4]: Because of the `for (const assetGroup of tokenBundle)` syntax. Eslint doesn't like that and I suppose it prefers `await Promise.awaitAll(tokenBundle.map(async ...))` (I didn't check the syntax is correct) but I think the order isn't guaranteed in that case so it wouldn't work for us. And AFAIK async foreach doesn't exist. The same is used in `cardanoSignTransaction.ts` btw.

I'll add a comment here (and in `cardanoSignTransaction.ts`) to make it clear.

File: packages/connect/src/types/api/cardanoSignTransaction.ts (hunk 5 above)
Comment (gabrielKerekes) [reply packages/connect/src/types/api/cardanoSignTransaction.ts:22]: Yes, indeed. Good points, I'll update the `.md` files, thanks 👍

File: packages/connect/src/api/cardano/cardanoOutputs.ts (hunk 4 above)
Comment (gabrielKerekes) [reply packages/connect/src/api/cardano/cardanoOutputs.ts:4]: Done in https://github.com/vacuumlabs/trezor-suite/pull/1/commits/58062096a0b3ab7b5e023ad1e11789368ceba375.

File: packages/connect/src/api/cardano/cardanoOutputs.ts (hunk 3 above)
Comment (gabrielKerekes) [reply packages/connect/src/api/cardano/cardanoOutputs.ts]: I added my suggestion in https://github.com/vacuumlabs/trezor-suite/pull/1/commits/58062096a0b3ab7b5e023ad1e11789368ceba375.

File: packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts (hunk 6 above)
Comment (gabrielKerekes) [reply packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts]: 👍 Fixed in https://github.com/vacuumlabs/trezor-suite/pull/1/commits/58062096a0b3ab7b5e023ad1e11789368ceba375.

File: packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts (hunk 8 above)
Comment (gabrielKerekes) [reply packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts]: Nice, I'm not sure how that happened. Thanks! Fixed in https://github.com/vacuumlabs/trezor-suite/pull/1/commits/58062096a0b3ab7b5e023ad1e11789368ceba375

File: packages/connect/src/types/api/cardanoSignTransaction.ts (hunk 5 above)
Comment (gabrielKerekes) [reply packages/connect/src/types/api/cardanoSignTransaction.ts:22]: Done in https://github.com/vacuumlabs/trezor-suite/pull/1/commits/6ef0d71261ccbf7424f20d36d7eef7635cfefab1.

File: packages/connect/src/api/cardanoSignTransaction.ts (hunk 1 above)
Comment (davidmisiak) [reply packages/connect/src/api/cardanoSignTransaction.ts:295]: Alright.

File: packages/connect/src/api/cardano/cardanoUtils.ts (hunk 2 above)
Comment (davidmisiak) [reply packages/connect/src/api/cardano/cardanoUtils.ts:13]: Ok, thanks.

File: packages/connect/src/api/cardano/cardanoOutputs.ts (hunk 3 above)
Comment (davidmisiak) [reply packages/connect/src/api/cardano/cardanoOutputs.ts]: Alright, your suggestion seems reasonable to me 👍

File: packages/connect/src/types/api/cardanoSignTransaction.ts (hunk 5 above)
Comment (davidmisiak) [reply packages/connect/src/types/api/cardanoSignTransaction.ts:22]: Cool, but shouldn't we also add the descriptions of the new Babbage tx items to `cardanoSignTransaction.md`?

Comment (gabrielKerekes) [reply packages/connect/src/types/api/cardanoSignTransaction.ts:22]: I think we should :D Sorry, I completely missed it. Updated. I squashed the commit though.

File: packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts (hunk 7 above)
Comment (gabrielKerekes) [reply packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts:417]: I added such a test - https://github.com/vacuumlabs/trezor-suite/pull/1/files#diff-6e989ad5e80587d676b0177bf290b661fc11290f47f3348fa1b1f16b6f9ed0e1R2398.

**Comment by davidmisiak** [comment]
Merged in https://github.com/trezor/trezor-suite/pull/5752
